helix-core/src/auto_pairs.rs

```diff
@@ -18,9 +20,98 @@ pub const PAIRS: &[(char, char)] = &[
     ('`', '`'),
 ];
 
-// [TODO] build this dynamically in language config. see #992
-const OPEN_BEFORE: &str = "([{'\":;,> \n\r\u{000B}\u{000C}\u{0085}\u{2028}\u{2029}";
-const CLOSE_BEFORE: &str = ")]}'\":;,> \n\r\u{000B}\u{000C}\u{0085}\u{2028}\u{2029}"; // includes space and newlines
+/// The type that represents the collection of auto pairs,
+/// keyed by the opener.
+#[derive(Debug, Clone)]
+pub struct AutoPairs(HashMap<char, Rc<AutoPair>>);
+
+/// Represents the config for a particular pairing.
+#[derive(Debug)]
+pub struct AutoPair {
+    pub open: char,
+    pub close: char,
+}
+
+impl AutoPair {
+    /// true if open == close
+    pub fn same(&self) -> bool {
+        self.open == self.close
+    }
+
+    /// true if all of the pair's conditions hold for the given document and range
+    pub fn should_close(&self, doc: &Rope, range: &Range) -> bool {
+        let mut should_close = Self::next_is_not_alpha(doc, range);
+
+        if self.same() {
+            should_close &= Self::prev_is_not_alpha(doc, range);
+        }
+
+        should_close
+    }
+
+    pub fn next_is_not_alpha(doc: &Rope, range: &Range) -> bool {
+        let cursor = range.cursor(doc.slice(..));
+        let next_char = doc.get_char(cursor);
+        next_char.map(|c| !c.is_alphanumeric()).unwrap_or(true)
+    }
+
+    pub fn prev_is_not_alpha(doc: &Rope, range: &Range) -> bool {
+        let cursor = range.cursor(doc.slice(..));
+        let prev_char = prev_char(doc, cursor);
+        prev_char.map(|c| !c.is_alphanumeric()).unwrap_or(true)
+    }
+}
+
+impl From<(char, char)> for AutoPair {
+    fn from(pair: (char, char)) -> Self {
+        Self {
+            open: pair.0,
+            close: pair.1,
+        }
+    }
+}
+
+impl From<(&char, &char)> for AutoPair {
+    fn from(pair: (&char, &char)) -> Self {
+        Self {
+            open: *pair.0,
+            close: *pair.1,
+        }
+    }
+}
```

Comment:
Is this ever necessary? chars are `Copy`

Reply:
Yeah, we're actually using this when we pass in the user mapping because `HashMap::iter` produces `(&char, &char)` and for some reason `cloned` doesn't work. There could be something I'm not realizing we could do instead.